### PR TITLE
Derive TryIntoJs for structs

### PR DESCRIPTION
**Pull request by mkawalec**
This implements https://github.com/infinyon/node-bindgen/issues/147, without the bonus points. 

I am fairly unhappy about the fact that we demand fields to be `Clone`. Because `try_into_js` takes ownership of `Self`, I am unsure how to run it for each of the struct's fields without cloning. Please do let me know if there is a way of making the borrow checker happy without cloning, it feels like there should be one.

Added a `TryIntoJs` instance for `usize` by casting it to `u64`. I don't think we expect this code to run on architectures wider than 64 bits, so the conversion should be a safe one.

### CODE REVIEW

**Comment by nicholastmosher** [comment]
Hi @mkawalec, thanks for the PR! I'll try to read through it and have some comments by the end of the day :)

File: nj-derive/ui-tests/fail_structure.rs [hunk 1]
```diff
@@ -1,9 +0,0 @@
-use node_bindgen::derive::node_bindgen;
-
- 
-#[node_bindgen]
-struct Example4{}
-
-fn main() {
-    
-}
```
Comment (sehz) [inline nj-derive/ui-tests/fail_structure.rs:9]: Parser doesn't support `Enum` and `Union`.  That should be include in fail cases

**Review by sehz — CHANGES_REQUESTED** [review]
Very impressive work.  Just one comments for now. I will have more feedback after digesting more.

File: nj-derive/ui-tests/fail_structure.rs (hunk 1 above)
Comment (mkawalec) [reply nj-derive/ui-tests/fail_structure.rs:9]: Good idea

**Comment by mkawalec** [comment]
@sehz Thank you. I didn't look into using `&self` in `try_to_js` too hard, but if there is no explicit reason why it must take the ownership, I can move it to a reference in the second half of the week.

**Comment by nicholastmosher** [comment]
I took a first-pass read through this today and everything looks really good, I think you did a great job. I wanted to play around with cargo-expand on some of the examples but for some reason it is not working for me, so I wasn't able to dig around with that like I was planning to. Tomorrow I'll take another deeper look and try to test the limits of this a bit more than the one test. Thanks again for putting in such great work!

**Comment by sehz** [comment]
Any update? Just need add fail test case

**Comment by mkawalec** [comment]
@sehz I haven't had any free time in the meantime, with a bit of luck this will happen this week. I want to get rid of the `clone` call on object fields before we merge this in.

**Comment by mkawalec** [comment]
I think I've got it - now only references need to be `Clone`, with no cloning happening for simple values.

**Comment by nicholastmosher** [comment]
@mkawalec I've just finished reading through this again and I think I'm happy with just about everything in here. The one question I have is why you decided to add support for unit structs? It seems to me that there isn't much use to converting those at all. It also took me awhile to figure out that they turned into empty objects. I was wondering if you could just talk me through why you decided to do it that way?

**Comment by mkawalec** [comment]
My reasoning was purely consistency based - if structs translate to objects, a unit struct would translate to an object with no keys. I was mostly thinking about allowing complex structs with many levels of structs embedded inside of them to convert automagically into Node.

That being said I am in no way set on that conversion - I can definitely see a merit of using a `null` to represent unit structs, or even a string equal to that struct's name. Do you have a more appropriate conversion in mind?

**Comment by nicholastmosher** [comment]
I suppose my initial reaction is that any of those options could make sense, but we currently have not seen a use-case that has demanded any approach over the others. With that being the case, I think I'd be more inclined to omit this functionality for now and wait to see whether anybody comes along who has a specific need for this and see what their requirements are. Would you mind removing the Unit struct support and perhaps we can open an issue to talk about its use case in the future?

**Comment by mkawalec** [comment]
Sure, that's also a valid approach. I'll remove the unit struct support today and replace it with an error message telling the user to reach out if they need it.

**Comment by nicholastmosher** [comment]
That sounds awesome, thanks!

**Review by nicholastmosher — APPROVED** [review]
Thanks for making those changes, I think everything looks good now! @sehz what do you think?

**Comment by sehz** [comment]
Update changelog and bump platform version

File: nj-core/src/convert.rs [hunk 2]
```diff
@@ -34,12 +34,19 @@ impl TryIntoJs for i32 {
         js_env.create_int32(self)
     }
 }
+
```
Comment (sehz) [inline nj-core/src/convert.rs:37]: Can you bump up version for `nj-core`?

**Review by sehz — CHANGES_REQUESTED** [review]
LGTM.  Couple of minor housekeeping:
* Bump up versions for effected crates
* Add to changelog
* CI failures

**Comment by sehz** [comment]
rebase as well

**Comment by mkawalec** [comment]
So the rebase went weirdly... It looks like it still does what it is supposed to do though, I highly recommend squash merge over rebasing for that exact reason.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,9 +2,22 @@
 
 ## Unreleased
 ### Improvements
-### Fixed
 
-## [4.2.2] - TBD
+## [4.4.0] - TBD
+### Improvements
```
Comment (sehz) [inline CHANGELOG.md:7]: looks like something went wrong with rebase?

**Review by sehz — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (mkawalec) [reply CHANGELOG.md:7]: Looks fine to me? But feel free to modify however you want and merge this.

**Review by sehz — APPROVED** [review]

**Comment by sehz** [comment]
Great work.  Thanks for contribution.

**Pull request merged**